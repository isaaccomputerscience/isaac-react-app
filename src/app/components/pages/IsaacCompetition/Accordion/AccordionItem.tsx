import React, { memo } from "react";
import { Card, CardBody, CardHeader, Collapse } from "reactstrap";

interface AccordionItemProps {
  id: string;
  title: string;
  section: string[];
  open: string | null;
  isLast: boolean;
  setOpenState: (id?: string) => void;
}

const renderSectionContent = (section: string[], isList: boolean) => (
  <>
    {section[0]}
    {isList ? (
      <ul>
        {section.slice(1, -1).map((item) => (
          <li key={item}>{item}</li>
        ))}
      </ul>
    ) : (
      section.slice(1, -1).map((item, index) => <p key={index}>{item}</p>)
    )}
    {section[section.length - 1]}
  </>
);

const AccordionItem = memo(({ id, title, section, open, isLast, setOpenState }: AccordionItemProps) => {
  const toggle = (id: string) => {
    setOpenState(open === id ? undefined : id);
  };

  const headerClasses = [
    "accordion-button p-3 m-0 d-flex justify-content-between",
    id === "0" && "rounded-top",
    isLast && "rounded-bottom border-bottom-0",
    open !== id && "border-bottom border-dark",
  ]
    .filter(Boolean)
    .join(" ");

  return (
    <Card eventKey={id}>
      <CardHeader className={headerClasses} onClick={() => toggle(id)}>
        {title}
        {open === id ? (
          <img src={"/assets/chevron_up.svg"} alt="Up Icon" className="p-2" />
        ) : (
          <img src={"/assets/chevron_down.svg"} alt="Down Icon" className="p-2" />
        )}
      </CardHeader>

      <Collapse isOpen={open === id}>
        <CardBody className="p-3 bg-white border-bottom border-dark">{renderSectionContent(section, true)}</CardBody>
      </Collapse>
    </Card>
  );
});

AccordionItem.displayName = "AccordionItem";

export default AccordionItem;
