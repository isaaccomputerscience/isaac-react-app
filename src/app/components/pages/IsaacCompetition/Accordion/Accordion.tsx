import React from "react";
import AccordionItem from "./AccordionItem";

interface AccordionProps {
  sections: {
    id: string;
    title: string;
    section: (string | string[])[];
  }[];
  open: string | null;
  setOpenState: (id: string | undefined) => void;
}

const Accordion = ({ sections, open, setOpenState }: AccordionProps) => {
  return (
    <div className="accordion accordion-body">
      {sections.map(({ id, title, section }, index) => (
        <AccordionItem
          key={id}
          id={id}
          title={title}
          section={section}
          open={open}
          isLast={index === sections.length - 1}
          setOpenState={setOpenState}
        />
      ))}
    </div>
  );
};

export default Accordion;
