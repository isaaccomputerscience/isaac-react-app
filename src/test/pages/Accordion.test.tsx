import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Accordion from "../../app/components/pages/IsaacCompetition/Accordion/Accordion";

describe("Accordion Component", () => {
  const sections = [
    {
      id: "1",
      title: "Section 1",
      section: ["Item 1", ["Sub-item 1", "Sub-item 2"]],
    },
    {
      id: "2",
      title: "Section 2",
      section: ["Item 3", "Item 3"],
    },
  ];
  const setOpenState = jest.fn();

  it("renders Accordion component", () => {
    render(<Accordion sections={sections} open={null} setOpenState={setOpenState} />);
    expect(screen.getByText("Section 1")).toBeInTheDocument();
    expect(screen.getByText("Section 2")).toBeInTheDocument();
  });

  it("renders correct number of AccordionItem components", () => {
    render(<Accordion sections={sections} open={null} setOpenState={setOpenState} />);
    const items = screen.getAllByTestId("accordion-item");
    expect(items.length).toBe(sections.length);
  });

  it("calls setOpenState with correct argument when AccordionItem is clicked", () => {
    render(<Accordion sections={sections} open={null} setOpenState={setOpenState} />);
    const firstItem = screen.getAllByText("Section 1")[0];
    fireEvent.click(firstItem);
    expect(setOpenState).toHaveBeenCalledWith("1");
  });

  it("renders the section content correctly when open", () => {
    render(<Accordion sections={sections} open="1" setOpenState={setOpenState} />);
    expect(screen.getAllByText("Item 1")[0]).toBeInTheDocument();
    expect(screen.getByText("Sub-item 1")).toBeInTheDocument();
    expect(screen.getByText("Sub-item 2")).toBeInTheDocument();
  });

  it("renders the section content as a list when it contains a sub-array", () => {
    render(<Accordion sections={sections} open="1" setOpenState={setOpenState} />);
    const listItems = screen.getAllByRole("listitem");
    expect(listItems.length).toBe(2);
    expect(listItems[0]).toHaveTextContent("Sub-item 1");
    expect(listItems[1]).toHaveTextContent("Sub-item 2");
  });

  it("matches the snapshot when closed", () => {
    const { asFragment } = render(<Accordion sections={sections} open={null} setOpenState={setOpenState} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders HTML content with links correctly", () => {
    const sectionsWithLinks = [
      {
        id: "1",
        title: "Section with Links",
        section: ["Visit <a href='https://example.com'>our website</a> for more info"],
      },
    ];
    render(<Accordion sections={sectionsWithLinks} open="1" setOpenState={setOpenState} />);

    const link = screen.getByRole("link", { name: /our website/i });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute("href", "https://example.com");
  });

  it("renders deeply nested lists correctly", () => {
    const sectionsWithNestedLists = [
      {
        id: "1",
        title: "Nested Lists",
        section: ["Main item", ["Level 1 item", ["Level 2 item", ["Level 3 item"]]]],
      },
    ];
    render(<Accordion sections={sectionsWithNestedLists} open="1" setOpenState={setOpenState} />);

    expect(screen.getByText("Main item")).toBeInTheDocument();
    expect(screen.getByText("Level 1 item")).toBeInTheDocument();
    expect(screen.getByText("Level 2 item")).toBeInTheDocument();
    expect(screen.getByText("Level 3 item")).toBeInTheDocument();

    // Check that nested lists are properly structured
    const lists = screen.getAllByRole("list");
    expect(lists.length).toBeGreaterThan(1);
  });

  it("shows correct chevron icons based on open state", () => {
    render(<Accordion sections={sections} open="1" setOpenState={setOpenState} />);

    // First section should show up chevron (open)
    const upIcon = screen.getByAltText("Up Icon");
    expect(upIcon).toBeInTheDocument();
    expect(upIcon).toHaveAttribute("src", "/assets/chevron_up.svg");

    // Second section should show down chevron (closed)
    const downIcon = screen.getByAltText("Down Icon");
    expect(downIcon).toBeInTheDocument();
    expect(downIcon).toHaveAttribute("src", "/assets/chevron_down.svg");
  });

  it("toggles accordion state correctly when clicked", () => {
    const { rerender } = render(<Accordion sections={sections} open={null} setOpenState={setOpenState} />);

    const firstItem = screen.getAllByText("Section 1")[0];

    fireEvent.click(firstItem);
    expect(setOpenState).toHaveBeenCalledWith("1");

    rerender(<Accordion sections={sections} open="1" setOpenState={setOpenState} />);

    fireEvent.click(firstItem);
    expect(setOpenState).toHaveBeenCalledWith(undefined);
  });

  it("handles empty sections gracefully", () => {
    const emptySections = [
      {
        id: "1",
        title: "Empty Section",
        section: [],
      },
    ];
    render(<Accordion sections={emptySections} open="1" setOpenState={setOpenState} />);

    expect(screen.getByText("Empty Section")).toBeInTheDocument();
    // Should not crash or throw errors
  });

  it("handles mixed content types correctly", () => {
    const mixedContentSections = [
      {
        id: "1",
        title: "Mixed Content",
        section: [
          "Plain text",
          ["List item 1", "List item 2"],
          "More plain text with <a href='#'>a link</a>",
          [["Deeply nested item"]],
        ],
      },
    ];
    render(<Accordion sections={mixedContentSections} open="1" setOpenState={setOpenState} />);

    expect(screen.getByText("Plain text")).toBeInTheDocument();
    expect(screen.getByText("List item 1")).toBeInTheDocument();
    expect(screen.getByText("More plain text with")).toBeInTheDocument();
    expect(screen.getByText("Deeply nested item")).toBeInTheDocument();
    expect(screen.getByRole("link")).toBeInTheDocument();
  });

  it("matches snapshot when open", () => {
    const { asFragment } = render(<Accordion sections={sections} open="1" setOpenState={setOpenState} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with nested content", () => {
    const nestedSections = [
      {
        id: "1",
        title: "Nested Section",
        section: ["Main content", ["Sub item 1", "Sub item 2"], "More content"],
      },
    ];
    const { asFragment } = render(<Accordion sections={nestedSections} open="1" setOpenState={setOpenState} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
