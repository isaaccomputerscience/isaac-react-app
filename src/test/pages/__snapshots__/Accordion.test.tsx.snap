// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component matches snapshot when open 1`] = `
<DocumentFragment>
  <div
    class="accordion accordion-body"
  >
    <div
      class="card"
      data-testid="accordion-item"
      eventkey="1"
    >
      <div
        class="accordion-button p-3 m-0 d-flex justify-content-between card-header"
      >
        Section 1
        <img
          alt="Up Icon"
          class="p-2"
          src="/assets/chevron_up.svg"
        />
      </div>
      <div
        class="collapse show"
      >
        <div
          class="p-3 bg-white border-bottom border-dark card-body"
        >
          <p>
            Item 1
          </p>
          <ul>
            <li>
              Sub-item 1
            </li>
            <li>
              Sub-item 2
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="card"
      data-testid="accordion-item"
      eventkey="2"
    >
      <div
        class="accordion-button p-3 m-0 d-flex justify-content-between rounded-bottom border-bottom-0 border-bottom border-dark card-header"
      >
        Section 2
        <img
          alt="Down Icon"
          class="p-2"
          src="/assets/chevron_down.svg"
        />
      </div>
      <div
        class="collapse"
      >
        <div
          class="p-3 bg-white border-bottom border-dark card-body"
        >
          <p>
            Item 3
          </p>
          <p>
            Item 3
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Accordion Component matches snapshot with nested content 1`] = `
<DocumentFragment>
  <div
    class="accordion accordion-body"
  >
    <div
      class="card"
      data-testid="accordion-item"
      eventkey="1"
    >
      <div
        class="accordion-button p-3 m-0 d-flex justify-content-between rounded-bottom border-bottom-0 card-header"
      >
        Nested Section
        <img
          alt="Up Icon"
          class="p-2"
          src="/assets/chevron_up.svg"
        />
      </div>
      <div
        class="collapse show"
      >
        <div
          class="p-3 bg-white border-bottom border-dark card-body"
        >
          <p>
            Main content
          </p>
          <ul>
            <li>
              Sub item 1
            </li>
            <li>
              Sub item 2
            </li>
          </ul>
          <p>
            More content
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Accordion Component matches the snapshot when closed 1`] = `
<DocumentFragment>
  <div
    class="accordion accordion-body"
  >
    <div
      class="card"
      data-testid="accordion-item"
      eventkey="1"
    >
      <div
        class="accordion-button p-3 m-0 d-flex justify-content-between border-bottom border-dark card-header"
      >
        Section 1
        <img
          alt="Down Icon"
          class="p-2"
          src="/assets/chevron_down.svg"
        />
      </div>
      <div
        class="collapse"
      >
        <div
          class="p-3 bg-white border-bottom border-dark card-body"
        >
          <p>
            Item 1
          </p>
          <ul>
            <li>
              Sub-item 1
            </li>
            <li>
              Sub-item 2
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="card"
      data-testid="accordion-item"
      eventkey="2"
    >
      <div
        class="accordion-button p-3 m-0 d-flex justify-content-between rounded-bottom border-bottom-0 border-bottom border-dark card-header"
      >
        Section 2
        <img
          alt="Down Icon"
          class="p-2"
          src="/assets/chevron_down.svg"
        />
      </div>
      <div
        class="collapse"
      >
        <div
          class="p-3 bg-white border-bottom border-dark card-body"
        >
          <p>
            Item 3
          </p>
          <p>
            Item 3
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
